File: go/deck.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"time"
)

type deck []card

type card struct {
	suit  string
	value string
}

func (d deck) print() {
	for i, c := range d {
		fmt.Println(i, c)
	}
}

func (d deck) toString() string {
	return fmt.Sprint(d)
}

func (d deck) saveToFile(filename string) error {
	return ioutil.WriteFile(filename, []byte(d.toString()), 0666)
}

func deckFromFile(filename string) deck {
	byteslice, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Println(err)
		os.Exit(1)
	}

	return deck()
}

func newDeck() deck {
	suits := []string{"♠️", "♦️", "♥️", "♣️"}
	vals := []string{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"}
	cards := deck{}

	for _, suit := range suits {
		for _, val := range vals {
			cards = append(cards, suit+val)
		}
	}

	return cards
}

func (d deck) shuffle() {
	rand.Seed(time.Now().UnixNano())
	for i := len(d) - 1; i >= 1; i -= 1 {
		j := rand.Intn(i)
		d[j], d[i] = d[i], d[j]
	}
}

func deal(d deck, handSize int) (deck, deck) {
	return d[:handSize], d[handSize:]
}

File: rust/src/deck.rs
use crate::card::{Card, Suit, Value};
use std::fmt;
use rand::thread_rng;
use rand::seq::SliceRandom;

#[derive(Debug)]
pub struct Deck {
    cards: Vec<Card>,
}

impl std::fmt::Display for Deck {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> std::fmt::Result {

        let mut output = String::new();
        for c in &self.cards {
            output += &format!("{}\n", c);
        }
        write!(f, "{}", output)
    }
}

impl Deck {
    pub fn new() -> Self {
        let mut cards = Vec::new();

        let suits = vec![Suit::Spade, Suit::Diamond, Suit::Heart, Suit::Club];

        let values = vec![
            Value::Ace,
            Value::Two,
            Value::Three,
            Value::Four,
            Value::Five,
            Value::Six,
            Value::Seven,
            Value::Eight,
            Value::Nine,
            Value::Ten,
            Value::Jack,
            Value::Queen,
            Value::King,
        ];

        for s in suits {
            for v in &values {
                cards.push(Card::new(s, *v));
            }
        }

        Deck { cards }
    }

    pub fn shuffle(&mut self){
        self.cards.shuffle(&mut thread_rng())
    }

    pub fn deal(&mut self, hand_size: usize)  {
        self.cards = self.cards.split_off(hand_size)
    }

    pub fn size(&self) -> usize{
        self.cards.len()
    }
}
